### ISSUE TITLE
API of DebugElement.attributes has changed in Ivy

### ISSUE THREAD

**Issue by sharikovvladislav** [angular/angular#33672]
# 🐞 bug report

### Affected Package

 The issue is caused by package angular/core

### Is this a regression?

 Yes, the previous version in which this bug was not present was: 9.0.0-rc.1 WITHOUT IVY.

### Description

There are two test outputs below. One is with Ivy enabled. One is with Ivy disabled. Same version: 9.0.0-rc.1.

Ivy disabled:
```
07 11 2019 16:45:38.713:INFO [Chrome 78.0.3904 (Mac OS X 10.14.6)]: Connected on socket wLxAErGHIZ52HWpWAAAA with id 69969590
LOG: '>> de.attributes', Object{ng-version: '9.0.0-rc.1', tabindex: null}
Chrome 78.0.3904 (Mac OS X 10.14.6): Executed 1 of 2 SUCCESS (0 secs / 0.062 secs)
Chrome 78.0.3904 (Mac OS X 10.14.6): Executed 2 of 2 SUCCESS (0.105 secs / 0.078 secs)
TOTAL: 2 SUCCESS
TOTAL: 2 SUCCESS
```

Ivy enabled:
```
07 11 2019 16:44:53.688:INFO [Chrome 78.0.3904 (Mac OS X 10.14.6)]: Connected on socket zalDmgVWKSP5geYKAAAA with id 8272420
LOG: '>> de.attributes', Object{id: 'root-ng-internal-isolated-1', ng-version: '9.0.0-rc.1'}
Chrome 78.0.3904 (Mac OS X 10.14.6): Executed 1 of 2 SUCCESS (0 secs / 0.037 secs)
Chrome 78.0.3904 (Mac OS X 10.14.6) AppComponent debug element FAILED
        Error: Expected undefined to be null. Tip: To check for deep equality, use .toEqual() instead of .toBe().
            at <Jasmine>
            at UserContext.<anonymous> (http://localhost:9877/_karma_webpack_/src/app/app.component.spec.ts:29:36)
            at ZoneDelegate.invoke (http://localhost:9877/_karma_webpack_/node_modules/zone.js/dist/zone-evergreen.js:365:1)
            at ProxyZoneSpec.onInvoke (http://localhost:9877/_karma_webpack_/node_modules/zone.js/dist/zone-testing.js:305:1)
Chrome 78.0.3904 (Mac OS X 10.14.6): Executed 2 of 2 (1 FAILED) (0 secs / 0.129 secs)
Chrome 78.0.3904 (Mac OS X 10.14.6) AppComponent debug element FAILED
        Error: Expected undefined to be null. Tip: To check for deep equality, use .toEqual() instead of .toBe().
            at <Jasmine>
            at UserContext.<anonymous> (http://localhost:9877/_karma_webpack_/src/app/app.component.spec.ts:29:36)
            at ZoneDelegate.invoke (http://localhost:9877/_karma_webpack_/node_modules/zone.js/dist/zone-evergreen.js:365:1)
Chrome 78.0.3904 (Mac OS X 10.14.6): Executed 2 of 2 (1 FAILED) (0.158 secs / 0.129 secs)
TOTAL: 1 FAILED, 1 SUCCESS
TOTAL: 1 FAILED, 1 SUCCESS
```

Same component, same tests.

The component:

```
@Component({
  selector: 'example-app',
  template: `
    <button (click)="toggle()">Toggle</button>
  `,
  host: {
    '[attr.tabindex]': 'disabled ? null : 0'
  },
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class AppComponent {
  disabled = false;

  toggle() {
    this.disabled = !this.disabled;
  }
}
```

The unit test:

```
  it('debug element', () => {
    const fixture = TestBed.createComponent(AppComponent);
    const app = fixture.debugElement.componentInstance;

    app.disabled = true;
    fixture.detectChanges();

    const de = fixture.debugElement;
    console.log('>> de.attributes', de.attributes);

    expect(de.attributes.tabindex).toBe(null);
  });
```

So, API of DebugElement.attributes has changed. 

Before Ivy: attribute was deleted -> null as attribute value.
After Ivy: attribute was deleted -> undefined as attribute value (no such attribute property in attributes).

Looks like it could be breaking change

## 🔬 Minimal Reproduction

1 Clone the repo https://github.com/sharikovvladislav/debug-element-api-changed-reprro
2 Go the the cloned repo dir
3 Install deps `yarn` 
4 Execute tests `yarn run test`

You will see the log like I gave above for *Ivy disabled* case.

Now, you can change `enableIvy` to true option in tsconfig . You will see *Ivy enabled* log. Test outputs for Ivy or non-Ivy are different. One test is failing with Ivy.

## 🌍  Your Environment

**Angular Version:**
<pre><code>

Angular CLI: 9.0.0-rc.0
Node: 10.16.0
OS: darwin x64
Angular: 9.0.0-rc.1
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-rc.0
@angular-devkit/build-angular     0.900.0-rc.0
@angular-devkit/build-optimizer   0.900.0-rc.0
@angular-devkit/build-webpack     0.900.0-rc.0
@angular-devkit/core              9.0.0-rc.0
@angular-devkit/schematics        9.0.0-rc.0
@angular/cli                      9.0.0-rc.0
@ngtools/webpack                  9.0.0-rc.0
@schematics/angular               9.0.0-rc.0
@schematics/update                0.900.0-rc.0
rxjs                              6.5.3
typescript                        3.6.4
webpack                           4.41.2
</code></pre>

## Questions
1. Is it known issue? I can't find any information
2. Will this be fixed or do you think it is normal behaviour? 

We have some failing unit tests because of this problem. I can fast-fix them, but still.

**Comment by alxhub** [comment]
@sharikovvladislav, it would be really helpful if you could provide the reproduction in the form of a stackblitz, although the code you've listed is at least complete and should allow us to investigate this issue, so thank you for that.

**Comment by sharikovvladislav** [comment]
Does Ivy works at stackblitz?

The problem is easily reproduced with unit tests. It (unit tests) is not
about the stackblitz. Or do you have an example how to do this?

I don't know how to show you the problem without units. ng API has changed
too. There are no more probe method. We could get DebugElement of any
element on the page through it. What method can do this now? Or how to get
DE from console in Ivy?

The problem is about DE, not about DOM. DOM renders correctly.


пт, 8 нояб. 2019 г. в 02:12, Alex Rickabaugh <notifications@github.com>:

**Comment by sharikovvladislav** [comment]
What do I mean?

My repro is about unit tests, since I don't know how to show the problem using dev build.

Here is what I see when my app using ViewEngine (using ng.probe): 

![image](https://user-images.githubusercontent.com/5622526/68467713-c593cf80-0227-11ea-88ea-4d5be371d01c.png)

In Ivy we don't have `ng.probe` method. Here is what ng contains in Ivy:

![image](https://user-images.githubusercontent.com/5622526/68467572-7e0d4380-0227-11ea-9210-13cb29498ee4.png)

How I can access debug element (what is analogue for `ng.probe($0)`) in Ivy? I think the problem is about DebugElement specifically. I know how to get DebugElement of something in ViewEngine, but I can't find a way to get DebugElement in Ivy.

Also, anyway I don't know how to run Ivy project in stackblitz. 

What I tried: 
- create a repo on github with next version of Angular + enableIvy in it and publish this repo to github. Here is my previous try: https://stackblitz.com/github/sharikovvladislav/angular-cdk-overlay-woot-with-ivy . Ivy is enabled, but we can see ngfactory files -> no Ivy enabled
- tried to find it and got no success . I found this: https://github.com/angular/angular-cli/issues/14353, somebody wrote, that Ivy works but it does not. Here is what we can see in their stackblitz example: 
![image](https://user-images.githubusercontent.com/5622526/68468533-46070000-0229-11ea-8ab1-69973402ad6e.png)

We can see enableIvy option, but there are *.ngfactory* files, so it is not about Ivy.

Can you, please, explain how to run Ivy Angular prooject on stackblitz?

**Comment by pkozlowski-opensource** [comment]
This was discussed in the team in the past and at the time I think that we've decided that ivy behaviour is correct since a removed attribute value is... well... `undefined`.

But I wonder if we shouldn't come back on the decision to be more consistent with the DOM Api. Try this in your favourite browser's console:

```
const div = document.createElement('div');
div.getAttribute('id'); => null
div.setAttribute('id', 'foo');
div.getAttribute('id'); => "foo"
div.removeAttribute('id');
div.getAttribute('id'); => null
```

My point here is that DOM APIs return `null` for both non-existing _and_ removed attributes so we should probably do the same in the debug element.

**Comment by sharikovvladislav** [comment]
Good point.

I think anything is accepted :) We have about 100~ units (out of 2.5k for Angular) red. Some of them are about this problem.

I just want to understand the situation :)

- if it is normal situation, that deleted attributes are undefined in DE now, I just rewrite some of our unit tests
- it it is not normal, I rewrite them any way, since we have strong desire to add 9.0.0 to production and I will just add comments about this possibly will be fixed in some future

One moment: if it is normal, looks like there are nothing in known issues / breaking changes about this change. Same code, same tests are now falling, so I think it is possible breaking change.

**Comment by sharikovvladislav** [comment]
By, the way. In other parts of Angular you have good abstractions. Probably, if there is some method like `attributes.isExists('name')`, you would have more flexibility to change your internal states.

**Comment by trotyl** [comment]
> This was discussed in the team in the past and at the time I think that we've decided that ivy behaviour is correct since a removed attribute value is... well... `undefined`.
> 
> But I wonder if we shouldn't come back on the decision to be more consistent with the DOM Api. Try this in your favourite browser's console:
> 
> ```
> const div = document.createElement('div');
> div.getAttribute('id'); => null
> div.setAttribute('id', 'foo');
> div.getAttribute('id'); => "foo"
> div.removeAttribute('id');
> div.getAttribute('id'); => null
> ```
> 
> My point here is that DOM APIs return `null` for both non-existing _and_ removed attributes so we should probably do the same in the debug element.

@pkozlowski-opensource That's not an equivalent API, [`Element` do have an `attributes` prop as well](https://developer.mozilla.org/en-US/docs/Web/API/Element/attributes), which provides the mapping of attribute names and values, as long as accessed via JavaScript property, then non-existing value would always got `undefined`.

<sub>Technically speaking "property not exist" and "property value is `undefined`" are still different scenarios.</sub>

**Comment by pkozlowski-opensource** [comment]
@trotyl I see, yes, this is another way of thinking about it! I wish all the DOM APIs would be self-consistent :-/ 

Anyway, I think that we need to pick one (`null` or `undefined`). The problem with the previous behaviour is that it was totally implementation dependent:
* asking for an attribute value that never existed would return `undefined`
* asking for an attribute value that was removed would return `null`
I don't think we can defend the above logic...

**Comment by pkozlowski-opensource** [comment]
@sharikovvladislav we've re-discussed this issue internally and decided that the ivy behaviour is correct and we are going to keep in v9 and going forward. 

In fact we need to realise that there are 2 issues being discussed here:
* consistent value returned for removed attributes and ones that never existed - we claim that the returned value should be the same in those 2 cases;
* `undefined` vs. `null` returned in above cases.

We now believe that it is a bug in VE to return different values for a removed attribute vs. an attribute that never existed - we should return the same value and we can see ivy's behaviour as a bug fix. Regarding `undefined` vs. `null` - let's settle on `undefined` as it is the same value that is returned when you try to access a non-existing property on an object.

tl;dr; is that you should update your test to test for both `undefined` and `null` (if you want to be compatible with both VE and ivy).

**Comment by sharikovvladislav** [comment]
Thank you!
